src/recovery/mod.rs

```diff
@@ -1,6 +1,7 @@
 mod errors;
 mod version;
 
+use anyhow::Context;
```

Comment:
as above. library code should use concrete error types, rather than type-erased errors.

Reply:
This project isn't a library. Useful error messages are more important in our logs.

Reply:
> This project isn't a library

in that case, why use the shared source directory? you're compiling a library target. I can think of a few good reasons to do this

- to overcome limitations that make it difficult to run integration tests against binary code
- in order to publish a subset of the codebase as a reusable library
- separation of concerns

the first doesn't apply- there are no integration tests. I don't think the second case applies, but if it did, it would be poor practice to use `thiserror`, since it should be up to the projects that *consume* the library to decide what to do with the concrete errors. Similarly for the third case- *presenting* errors would be a concern of the application target, not the binary target.

> Useful error messages are more important in our logs.

that's fine. i guess I would point out that using concrete error types in your library code doesn't preclude you from pretty printing the errors downstream in the application code (though it may well make it more verbose!). I'm *definitely* being overly pedantic here, but I guess i'm just trying to point out that you're using two, quite different error-handling strategies interchangeably in the same compilation target.

Reply:
I sound like some pedant who just wandered in off the street ;)

Feel free to go ahead and ignore all this

Reply:
> in that case, why use the shared source directory

That is because the pop-upgrade project has many components. The majority of the product is the daemon, but there's also a command line frontend, notification service, and a GTK widget to integrate into GNOME Settings. The GTK widget is its own sub-project that leans on some of the same code and types as the daemon and CLI frontend.

> I guess i'm just trying to point out that you're using two, quite different error-handling strategies interchangeably in the same compilation target.

That's what happens when you have a large project that's been developed over a decent period of time. The error handling strategy in pop-upgrade has changed as better error handling strategies in Rust have been developed. The newer code is written to take advantage of anyhow::Context, while the original code is still using thiserror. The old code isn't broken, and it hasn't been a priority to redo everything for the sake of doing away with thiserror.

Reply:
> That is because the pop-upgrade project has many components. 

why does the number of components determine whether the crate root is a `lib.rs`, a `main.rs`, or both? i'm not sure i understand the motivation for having both in your case. what value are you getting from the library target (compared to, say, a single binary target with submodules?

>The newer code is written to take advantage of anyhow::Context, while the original code is still using thiserror.

`anyhow` isn't newer. they were released at the same time, by the same author, for different use-cases.

Reply:
> anyhow isn't newer. they were released at the same time, by the same author, for different use-cases.

Feel free to go 2 years in the past of our commit history. This project began development a full year before thiserror and anyhow existed. The concept of deriving errors from enums existed before thiserror was released. The thiserror crate was a simple drop-in replacement for err-derive.

> why does the number of components determine whether the crate root is a lib.rs, a main.rs

The GTK widget generates a lib.so file to import into a C application. A lib.rs is necessary to import pop-upgrade into the GTK sub-project. Feel free to investigate our Makefile to see how everything is packaged.

Reply:
Ah that makes sense then. Thanks for the taking the time to explain, I appreciate it